File: CS/DXApplication9/CharacterPropertiesObject.cs
// Developer Express Code Central Example:
// How to implement the basic idea of the Microsoft Word "Format Painter" feature for RichEditControl
// 
// This example demonstrates how to copy the characters and paragraphs properties
// and apply formatting to the selected text. Try the Format Painter button on the
// ribbon Home tab.
// 
// To obtain the selected text range, the
// RichEditDocument.Document.Selection property is used. The characters and
// paragraphs properties are obtained in the BarCheckItem.CheckedChanged event
// handler. Then, they are stored in the CharacterPropertiesObject and
// ParagraphPropertiesObject object containers.
// 
// In order to change the current
// RichEditControl cursor, a MouseCursorCalculator descendant is implemented. Check
// the custom MouseCursorCalculator class Calculate method for clarification.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E5223

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Model;

namespace DXApplication9
{
    public class CharacterPropertiesObject : DevExpress.XtraRichEdit.API.Native.CharacterProperties
    {

        public CharacterPropertiesObject()
        {

        }

        public bool? AllCaps { get; set; }

        public Color? BackColor { get; set; }

        public bool? Bold { get; set; }

        public string FontName { get; set; }

        public float? FontSize { get; set; }

        public Color? ForeColor { get; set; }

        public bool? Hidden { get; set; }

        public bool? Italic { get; set; }

        public LangInfo? Language { get; set; }

        public bool? NoProof { get; set; }

        public void Reset()
        {
            throw new NotImplementedException();
        }
        public void Reset(CharacterPropertiesMask prop) {
            throw new NotImplementedException();
        }

        public DevExpress.XtraRichEdit.API.Native.StrikeoutType? Strikeout { get; set; }

        private bool? _Superscript;
        private bool? _Subscript;
        public bool? Subscript {
            get {
                return _Subscript;
            }
            set {
                if(!Convert.ToBoolean(_Superscript))
                    _Subscript = value;
            }
        }

        public bool? Superscript {
            get {
                return _Superscript;
            }
            set {
                if(!Convert.ToBoolean(_Subscript))
                    _Superscript = value;
            }
        }

        public DevExpress.XtraRichEdit.API.Native.UnderlineType? Underline { get; set; }

        public Color? UnderlineColor { get; set; }

        public DevExpress.XtraRichEdit.API.Native.CharacterStyle Style { get; set; }




     
    }
}

File: VB/DXApplication9/CharacterPropertiesObject.vb
' Developer Express Code Central Example:
' How to implement the basic idea of the Microsoft Word "Format Painter" feature for RichEditControl
' 
' This example demonstrates how to copy the characters and paragraphs properties
' and apply formatting to the selected text. Try the Format Painter button on the
' ribbon Home tab.
' 
' To obtain the selected text range, the
' RichEditDocument.Document.Selection property is used. The characters and
' paragraphs properties are obtained in the BarCheckItem.CheckedChanged event
' handler. Then, they are stored in the CharacterPropertiesObject and
' ParagraphPropertiesObject object containers.
' 
' In order to change the current
' RichEditControl cursor, a MouseCursorCalculator descendant is implemented. Check
' the custom MouseCursorCalculator class Calculate method for clarification.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E5223


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Model

Namespace DXApplication9
	Public Class CharacterPropertiesObject
        Implements DevExpress.XtraRichEdit.API.Native.CharacterProperties

		Public Sub New()

		End Sub

        Public Property AllCaps() As Boolean? Implements CharacterPropertiesBase.AllCaps

        Public Property BackColor() As Color? Implements CharacterPropertiesBase.BackColor

        Public Property Bold() As Boolean? Implements CharacterPropertiesBase.Bold

        Public Property FontName() As String Implements CharacterPropertiesBase.FontName

        Public Property FontSize() As Single? Implements CharacterPropertiesBase.FontSize

        Public Property ForeColor() As Color? Implements CharacterPropertiesBase.ForeColor

        Public Property Hidden() As Boolean? Implements CharacterPropertiesBase.Hidden

        Public Property Italic() As Boolean? Implements CharacterPropertiesBase.Italic

        Public Property Language() As LangInfo? Implements CharacterPropertiesBase.Language

        Public Property NoProof() As Boolean? Implements CharacterPropertiesBase.NoProof

        Public Sub Reset() Implements CharacterPropertiesBase.Reset
            Throw New NotImplementedException()
        End Sub
        Public Sub Reset(ByVal prop As CharacterPropertiesMask) Implements CharacterPropertiesBase.Reset
            Throw New NotImplementedException()
        End Sub


        Private privateStrikeout? As DevExpress.XtraRichEdit.API.Native.StrikeoutType
        Public Property Strikeout() As DevExpress.XtraRichEdit.API.Native.StrikeoutType? Implements CharacterPropertiesBase.Strikeout
            Get
                Return privateStrikeout
            End Get
            Set(ByVal value As DevExpress.XtraRichEdit.API.Native.StrikeoutType?)
                privateStrikeout = value
            End Set
        End Property

		Private _Superscript? As Boolean
        Private _Subscript? As Boolean
        Public Property Subscript() As Boolean? Implements CharacterPropertiesBase.Subscript
            Get
                Return _Subscript
            End Get
            Set(ByVal value As Boolean?)
                If (Not Convert.ToBoolean(_Superscript)) Then
                    _Subscript = value
                End If
            End Set
        End Property

        Public Property Superscript() As Boolean? Implements CharacterPropertiesBase.Superscript
            Get
                Return _Superscript
            End Get
            Set(ByVal value As Boolean?)
                If (Not Convert.ToBoolean(_Subscript)) Then
                    _Superscript = value
                End If
            End Set
        End Property

        Public Property Underline() As DevExpress.XtraRichEdit.API.Native.UnderlineType? Implements CharacterPropertiesBase.Underline

        Public Property UnderlineColor() As Color? Implements CharacterPropertiesBase.UnderlineColor

        Public Property Style() As DevExpress.XtraRichEdit.API.Native.CharacterStyle Implements DevExpress.XtraRichEdit.API.Native.CharacterProperties.Style
    End Class
End Namespace
